sodetlib/operations/bias_dets.py

```diff
@@ -201,3 +201,408 @@ def bias_to_rfrac(S, cfg, rfrac=0.5, bias_groups=None, iva=None,
 
     return biases
 
+def biasstep_rebias(
+        S, cfg, target_percentage_rn = 0.5, bias_groups=None,
+        show_plots=True, make_plots=True):
+
+    """
+    re-bias detectors by taking bias_steps.
+    
+     Args
+    -----
+    S : SmurfControl
+        Pysmurf instance
+    cfg : DetConfig
+        det config instance
+    target_percentage_rn : float
+        this is the target percentage rn for the script to bias to, numbers should be
+        between 0.3 and 0.7 for general usage.
+    bias_groups : list
+        List of bias groups to bias. If None, this will default to the active
+        bgs specified in the device cfg.
+    make_plots : bool
+        If this is set to False, the script will still apply the bias-voltage but will not
+        generate plots
+    show_plots : bool
+        If this is set to False, the script will not show plots
+
+    Returns
+    ----------
+    bsa_final :
+        analysis result of the last biasstep taken with the re-biased bais voltage
+    vbias_estimate_final :
+        final bias_voltages calculated and applied by the script
+    """
+    if bias_groups is None:
+        bias_groups = cfg.dev.exp['active_bgs']
+    bias_groups = np.atleast_1d(bias_groups)
+
+    ## take the initial biasstep
+    S.log("taking the first biasstep")
+    S.log(f"Initial dc biases {S.get_tes_bias_bipolar_array()}")
+    bsa_0 = bias_steps.take_bias_steps(S, cfg)
+    if make_plots:
+        fig, ax = bias_steps.plot_Rfrac(bsa_0)
+        fname = sdl.make_filename(S, 'intial_Rfrac.png', plot=True)
+        fig.savefig(fname)
+        S.pub.register_file(fname, 'biasstep_rfrac', format='png', plot=True)
+        if not show_plots:
+            plt.close(fig)
+            
+            
+    ## load IV analysis result so can use dynamic step size
+    iva = iv.IVAnalysis.load(bsa_0.meta['iv_file'])
+                
+    repeat_biasstep = False
+    ## check for script failue
+    for bl in bias_groups:
+        mask_bg = np.where(bsa_0.bgmap==bl)[0]
+        per_bl_R0 = bsa_0.R0[mask_bg]
+        if np.isnan(np.nanmedian(per_bl_R0)):
+            ## the yield is very low, could it be the biasstep analysis script failed? repeat again
+            repeat_biasstep = True
+            S.log("the biasstep script might have failed, going to repeat the measurement")
+
+    if repeat_biasstep:
+        S.log("retaking biasstep due to failure")
+        bsa_0 = bias_steps.take_bias_steps(S, cfg)
+        repeat_biasstep = False
```

Comment:
I don't really like that this is forcing a retake of the bias steps, especially since if it fails a second time it will just continue on as if everything's ok. I think instead if the first bias steps fails, this should raise an exception with a useful error message.

Do you know how often this fails / why? We should try to fix in the bias step function directly instead of building fixes into this function.

Reply:
bias step failure has been noticed in multiple places including at the site in LATR, in Chicago LATRt, TSAT and Cornell. One of the failure mode might due to the if statement here: https://github.com/simonsobs/sodetlib/blob/4aafb94bb36cac85b754b0a1e28575d33e1b6243/sodetlib/operations/bias_steps.py#L645
Tho repeating the biasstep is not ideal, it is a working method given the current circumstances.

Reply:
but simply repeating it worked. yes Yaqiong and I have been reviewing some functions together.In my opinion I think the if statement can just be removed as we don't really rely on a accurate tes resistance from the biasstep when TES is not in-transition.

Reply:
Do you think it's mostly failing when a large fraction of detectors are superconducting? I notice you don't have retries for other times you call bias steps.

Reply:
I added in a retry for each bias step I think. I think the failing can happen either when a large fraction of detectors are superconducting or being normal. 
also repeating biasstep twice is like a old voodoo from my ACT days. I think such fix might make sense if one imagines some ADC needs to be 'waken'. I am not sure but I agree this can be redundant and can be removed after better understanding later.

Reply:
<img width="484" alt="Screenshot 2023-06-20 at 6 57 13 PM" src="https://github.com/simonsobs/sodetlib/assets/38056285/f0bee862-f8ae-49c1-a4cb-70056362c0fd">
<img width="1184" alt="Screenshot 2023-06-20 at 6 57 38 PM" src="https://github.com/simonsobs/sodetlib/assets/38056285/f424af4f-00c3-4451-87d7-7a350405c3eb">
this is one of the failure case I just found
after taking another bias_step this is fixed 
this is taken in LATRt, on Uv31, I believe the session id is 1687298049

Reply:
failure:
<img width="503" alt="Screenshot 2023-06-20 at 7 00 48 PM" src="https://github.com/simonsobs/sodetlib/assets/38056285/09abf
<img width="1209" alt="Screenshot 2023-06-20 at 7 01 01 PM" src="https://github.com/simonsobs/sodetlib/assets/38056285/0
<img width="506" alt="Screenshot 2023-06-20 at 7 01 25 PM" src="https://github.com/simonsobs/sodetlib/assets/38056285/31d00c17-10c1-47f7-be86-6b7ad439caab">
279dfe8-53f4-48de-b175-ec6991953072">
d19-1483-4a48-8419-e01e49fb9f56">
after repeating
<img width="482" alt="Screenshot 2023-06-20 at 7 01 39 PM" src="https://github.com/simonsobs/sodetlib/assets/38056285/5bac2819-f838-4dac-8b3b-95dee242f6a0">
<img width="542" alt="Screenshot 2023-06-20 at 7 01 52 PM" src="https://github.com/simonsobs/sodetlib/assets/38056285/349cca3f-bbc2-4afa-8342-272055e01fbe">
<img width="1199" alt="Screenshot 2023-06-20 at 7 02 01 PM" src="https://github.com/simonsobs/sodetlib/assets/38056285/03b76a19-c742-49d4-bc17-5c50de4d6365">

Reply:
awesome, that's really interesting. are these files all on simons1? I'll take a look tomorrow.